sparse_strips/vello_hybrid/examples/webgl/src/lib.rs

```diff
@@ -101,9 +111,378 @@ impl RendererWrapper {
+        surface_texture.present();
+
+        self.need_render = false;
+    }
+
+    fn resize(&mut self, width: u32, height: u32) {
+        self.canvas.set_width(width);
+        self.canvas.set_height(height);
+        self.width = width;
+        self.height = height;
+
+        self.scene = vello_hybrid::Scene::new(width as u16, height as u16);
+        self.renderer_wrapper.resize(width, height);
+
+        self.need_render = true;
+    }
+
+    fn next_scene(&mut self) {
+        self.current_scene = (self.current_scene + 1) % self.scenes.len();
+        self.transform = Affine::IDENTITY;
+        self.need_render = true;
+    }
+
+    fn prev_scene(&mut self) {
+        self.current_scene = if self.current_scene == 0 {
+            self.scenes.len() - 1
+        } else {
+            self.current_scene - 1
+        };
+        self.transform = Affine::IDENTITY;
+        self.need_render = true;
+    }
+
+    fn reset_transform(&mut self) {
+        self.transform = Affine::IDENTITY;
+        self.need_render = true;
+    }
+
+    fn handle_mouse_down(&mut self, x: f64, y: f64) {
+        self.mouse_down = true;
+        self.last_cursor_position = Some(Vec2::new(x, y));
+    }
+
+    fn handle_mouse_up(&mut self) {
+        self.mouse_down = false;
+        self.last_cursor_position = None;
+    }
+
+    fn handle_mouse_move(&mut self, x: f64, y: f64) {
+        let current_pos = Vec2::new(x, y);
+
+        if self.mouse_down {
+            if let Some(last_pos) = self.last_cursor_position {
+                let delta = current_pos - last_pos;
+                self.transform = Affine::translate(delta) * self.transform;
+                self.need_render = true;
+            }
+        }
+
+        self.last_cursor_position = Some(current_pos);
+    }
+
+    fn handle_wheel(&mut self, delta_y: f64) {
+        const ZOOM_STEP: f64 = 0.1;
+
+        if let Some(cursor_pos) = self.last_cursor_position {
+            let zoom_factor = (1.0 + delta_y * ZOOM_STEP).max(0.1);
+
+            // Zoom centered at cursor position
+            self.transform = Affine::translate(cursor_pos)
+                * Affine::scale(zoom_factor)
+                * Affine::translate(-cursor_pos)
+                * self.transform;
+
+            self.need_render = true;
+        } else {
+            // If no cursor position is known, zoom centered on screen
+            let center = Vec2::new(self.width as f64 / 2.0, self.height as f64 / 2.0);
+
+            let zoom_factor = (1.0 + delta_y * ZOOM_STEP).max(0.1);
+
+            self.transform = Affine::translate(center)
+                * Affine::scale(zoom_factor)
+                * Affine::translate(-center)
+                * self.transform;
+
+            self.need_render = true;
+        }
+    }
+}
+
+#[cfg(target_arch = "wasm32")]
+#[wasm_bindgen]
+extern "C" {
+    #[wasm_bindgen(js_name = requestAnimationFrame)]
+    fn request_animation_frame(f: &Closure<dyn FnMut()>);
```

Comment:
In release mode, the webgl svg (tiger) example takes around ~800ms to rerender. Is it the same on your end?

Reply:
Yes 😅 . When I profiled it, IIRC, most of the time was spent in stroke expansion. I think it's missing the fast stroke expansion work Raph is working on. See: [#vello > CPU sparse strip rendering to pixels @ 💬](https://xi.zulipchat.com/#narrow/channel/197075-vello/topic/CPU.20sparse.20strip.20rendering.20to.20pixels/near/508597866).

I think web doesn't get autovectorisation which is a separate issue entirely though 😅